File: python/copass-core/src/copass_core/resources/integrations.py
"""Integrations resource — provider-neutral OAuth → DataSource flow.

Port of ``typescript/packages/core/src/resources/integrations.ts``.

Phase 1A scope (read-only): ``catalog``, ``list``, plus the new
``list_accounts`` method that targets the
``GET /sandboxes/{id}/sources/integrations/accounts`` route added in
Phase 1A. Write methods (``connect``, ``disconnect``, ``reconcile``)
are included for parity with the TS surface — they're thin wrappers
over existing endpoints. Spec-driven write tools land in Phase 2.

The backing tool-source provider is server-side config; the SDK sees
a single unified surface. OAuth tokens stay with the provider — Copass
never persists them.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from copass_core.resources.base import BaseResource


def _base(sandbox_id: str) -> str:
    return f"/api/v1/storage/sandboxes/{sandbox_id}/sources/integrations"


class IntegrationsResource(BaseResource):
    """``/api/v1/storage/sandboxes/{id}/sources/integrations``.

    Browse available apps via :meth:`catalog`, list active connections
    via :meth:`list`, list raw upstream OAuth grants via
    :meth:`list_accounts`, start an OAuth flow via :meth:`connect`,
    and reconcile / disconnect for safety-net operations.
    """

    async def catalog(
        self,
        sandbox_id: str,
        *,
        q: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List or search the app catalog for a sandbox.

        Apps with curated ``pull_tool_calls`` defaults are tagged
        ``supported``; only those can be passed to :meth:`connect`.
        """
        query: Dict[str, Any] = {}
        if q is not None:
            query["q"] = q
        if limit is not None:
            query["limit"] = str(limit)
        if cursor is not None:
            query["cursor"] = cursor
        return await self._get(
            f"{_base(sandbox_id)}/catalog",
            query=query or None,
        )

    async def list_accounts(
        self,
        sandbox_id: str,
        *,
        app_slug: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List the user's raw OAuth accounts upstream.

        Distinct from :meth:`list`: connections are local DataSource
        rows materialised by the connect webhook; accounts are the raw
        grants on the upstream provider. Backs the ``list_connected_accounts``
        management tool.
        """
        query: Dict[str, Any] = {}
        if app_slug is not None:
            query["app_slug"] = app_slug
        return await self._get(
            f"{_base(sandbox_id)}/accounts",
            query=query or None,
        )

    async def connect(
        self,
        sandbox_id: str,
        app: str,
        *,
        success_redirect_uri: str,
        error_redirect_uri: str,
        scope: Optional[str] = None,
        project_id: Optional[str] = None,
        webhook_uri: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Mint a provider Connect URL for the given app.

        The returned ``connect_url`` is a provider-hosted OAuth page.
        The provider calls Copass's webhook on completion, creating a
        DataSource asynchronously.

        ``webhook_uri`` is an optional override forwarded verbatim to
        the upstream provider; webhook-using providers honor it,
        providers that do not use webhooks ignore it. Defaults to a
        URL composed server-side from the deployment's public base
        URL.
        """
        body: Dict[str, Any] = {
            "success_redirect_uri": success_redirect_uri,
            "error_redirect_uri": error_redirect_uri,
        }
        if scope is not None:
            body["scope"] = scope
        if project_id is not None:
            body["project_id"] = project_id
        if webhook_uri is not None:
            body["webhook_uri"] = webhook_uri
        return await self._post(f"{_base(sandbox_id)}/{app}/connect", body)

    async def list(
        self,
        sandbox_id: str,
        *,
        app: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List active integration-sourced connections in the sandbox.

        Optionally filter by ``app`` slug.
        """
        query: Dict[str, Any] = {}
        if app is not None:
            query["app"] = app
        return await self._get(
            f"{_base(sandbox_id)}/connections",
            query=query or None,
        )

    async def disconnect(self, sandbox_id: str, source_id: str) -> None:
        """Revoke the provider account (best-effort) and archive the
        DataSource locally. Idempotent in both directions.
        """
        await self._delete(f"{_base(sandbox_id)}/connections/{source_id}")

    async def reconcile(
        self,
        sandbox_id: str,
        *,
        scope: Optional[str] = None,
        project_id: Optional[str] = None,
        app: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Idempotently backfill missing DataSources from the provider.

        Safety net for dropped webhook deliveries. Always returns the
        post-reconcile state.
        """
        body: Dict[str, Any] = {}
        if scope is not None:
            body["scope"] = scope
        if project_id is not None:
            body["project_id"] = project_id
        if app is not None:
            body["app"] = app
        return await self._post(f"{_base(sandbox_id)}/reconcile", body)


__all__ = ["IntegrationsResource"]

File: typescript/packages/core/src/resources/integrations.ts
import { BaseResource } from './base.js';
import type {
  AppCatalogResponse,
  CatalogOptions,
  ConnectionsListResponse,
  ConnectRequest,
  ConnectResponse,
  IntegrationAccountListResponse,
  ListAccountsOptions,
  ListConnectionsOptions,
  ReconcileRequest,
  ReconcileResponse,
} from '../types/integrations.js';

const base = (sandboxId: string) =>
  `/api/v1/storage/sandboxes/${sandboxId}/sources/integrations`;

/**
 * Integrations resource — provider-neutral OAuth → DataSource flow.
 *
 * Browse available apps via {@link catalog}, start an OAuth flow
 * via {@link connect} (returns a hosted URL to hand to the user's
 * browser), and manage active connections via {@link list} /
 * {@link disconnect}.
 *
 * The backing integration provider is a server-side config; clients
 * see a single unified surface. OAuth tokens are held by the provider
 * — Copass never persists them.
 */
export class IntegrationsResource extends BaseResource {
  /** List or search the app catalog for a sandbox. Apps with
   * curated `pull_tool_calls` defaults are marked `supported`; only
   * those can be passed to {@link connect}. */
  async catalog(
    sandboxId: string,
    options: CatalogOptions = {},
  ): Promise<AppCatalogResponse> {
    const params = new URLSearchParams();
    if (options.q) params.set('q', options.q);
    if (options.limit) params.set('limit', String(options.limit));
    if (options.cursor) params.set('cursor', options.cursor);
    const qs = params.toString();
    const path = qs ? `${base(sandboxId)}/catalog?${qs}` : `${base(sandboxId)}/catalog`;
    return this.get<AppCatalogResponse>(path);
  }

  /** Mint a provider Connect URL for the given app + scope. The
   * returned `connect_url` is a provider-hosted page the user must
   * visit in a browser to complete OAuth. The provider calls
   * Copass's webhook on completion, creating a DataSource asynchronously. */
  async connect(
    sandboxId: string,
    app: string,
    request: ConnectRequest,
  ): Promise<ConnectResponse> {
    return this.post<ConnectResponse>(`${base(sandboxId)}/${app}/connect`, request);
  }

  /** List the user's raw upstream OAuth accounts.
   *
   * Distinct from {@link list}: connections are local DataSource rows
   * the connect webhook materialises; accounts are the raw grants on
   * the upstream provider. Backs the `list_connected_accounts`
   * management tool. Server-side at
   * `GET /sources/integrations/accounts`. */
  async listAccounts(
    sandboxId: string,
    options: ListAccountsOptions = {},
  ): Promise<IntegrationAccountListResponse> {
    const params = new URLSearchParams();
    if (options.app_slug) params.set('app_slug', options.app_slug);
    const qs = params.toString();
    const path = qs
      ? `${base(sandboxId)}/accounts?${qs}`
      : `${base(sandboxId)}/accounts`;
    return this.get<IntegrationAccountListResponse>(path);
  }

  /** List the sandbox's active integration-sourced connections.
   * Optionally filter by `app` slug. */
  async list(
    sandboxId: string,
    options: ListConnectionsOptions = {},
  ): Promise<ConnectionsListResponse> {
    const params = new URLSearchParams();
    if (options.app) params.set('app', options.app);
    const qs = params.toString();
    const path = qs
      ? `${base(sandboxId)}/connections?${qs}`
      : `${base(sandboxId)}/connections`;
    return this.get<ConnectionsListResponse>(path);
  }

  /** Revoke the provider account (best-effort) and archive the
   * DataSource locally. Idempotent in both directions. */
  async disconnect(sandboxId: string, sourceId: string): Promise<void> {
    await this.delete(`${base(sandboxId)}/connections/${sourceId}`);
  }

  /** Query the provider for the caller's connected accounts and
   * idempotently backfill missing DataSources. Safety net for dropped
   * webhook deliveries — call during a poll loop after
   * {@link connect} so a missed CONNECTION_SUCCESS doesn't leave the
   * user stuck. Always returns the post-reconcile state. */
  async reconcile(
    sandboxId: string,
    request: ReconcileRequest = {},
  ): Promise<ReconcileResponse> {
    return this.post<ReconcileResponse>(`${base(sandboxId)}/reconcile`, request);
  }
}
